File: miscellaneous/matrix-bfs-traversal/20230107/matrix_bfs.py
from typing import Tuple, List


class Solution:
    _offsets = (-1, 0), (0, 1), (1, 0), (0, -1)
    _matrix: Tuple[Tuple[int, int, int, int],
            Tuple[int, int, int, int],
            Tuple[int, int, int, int],
            Tuple[int, int, int, int],
    ]

    def matrix_bfs(self, matrix: Tuple[Tuple[int, int, int, int]]) -> List[int]:
        values: List[int] = []
        queue = [(0, 0)]
        seen = {(0, 0)}
        self._matrix = matrix

        while queue:
            row, col = queue.pop(0)
            values.append(self._matrix[row][col])

            for yOffset, xOffset in self._offsets:
                new_row = row + yOffset
                new_col = col + xOffset

                if (self._in_bounds(new_row, new_col) and
                        not (new_row, new_col) in seen):
                    seen.add((new_row, new_col))
                    queue.append((new_row, new_col))

        return values

    def _in_bounds(self, row: int, col: int) -> bool:
        return 0 <= row < len(self._matrix) and 0 <= col < len(self._matrix[0])

File: miscellaneous/matrix-bfs-traversal/20230107/MatrixBfs.php
<?php

declare(strict_types=1);

class MatrixBfs
{
    protected const DIRS = [[-1, 0], [0, 1], [1, 0], [0, -1]];
    protected array $matrix;
    protected array $seen;

    public function matrixBfs(array &$matrix): array
    {
        $this->matrix = $matrix;
        $values = [];
        $queue = [[0, 0]];
        $this->seen[serialize([0, 0])] = true;

        while ($queue) {
            [$row, $col] = array_shift($queue);
            $values[] = $matrix[$row][$col];

            foreach (static::DIRS as [$yOffset, $xOffset]) {
                $newRow = $row + $yOffset;
                $newCol = $col + $xOffset;

                if ($this->inBounds(row: $newRow, col: $newCol)
                        && !$this->hasSeen($newRow, $newCol)) {
                    $this->seen[serialize([$newRow, $newCol])] = true;
                    $queue[] = [$newRow, $newCol];
                }
            }
        }

        return $values;
    }

    protected function inBounds(int $row, int $col): bool
    {
        return 0 <= $row && $row < count($this->matrix)
            && 0 <= $col && $col < count($this->matrix[0]);
    }

    protected function hasSeen(int $row, int $col): bool
    {
        return array_key_exists(serialize([$row, $col]), $this->seen);
    }
}